### PR TITLE
Fixes #11610: file_copy_with_check

### PR DESCRIPTION
**Pull request by amousset**
https://www.rudder-project.org/redmine/issues/11610

### CODE REVIEW

**Comment by amousset** [comment]
I'll add a test with a different return code.

**Comment by amousset** [comment]
Commit modified

File: tree/30_generic_methods/file_copy_from_local_source_with_check.cf [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      "promisers"          slist => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
+      "class_prefix"      string => canonify(join("_", "promisers"));
+      "args"              slist => { "${source}", "${destination}", "${check_command}" };
+      "canonified_file"   string => canonify("${destination}");
+
+  defaults:
+      "rc_ok"             string => "0", if_match_regex => "";
+
+  classes:
+      "copy_needed" not => returnszero("diff \"${source}\" \"${destination}\"", "useshell");
```
Comment (ncharles) [inline tree/30_generic_methods/file_copy_from_local_source_with_check.cf]: could you add an "agent:" class here, to make sure it's not run with cf-promises ?

File: tree/30_generic_methods/file_copy_from_local_source_with_check.cf [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+      "pass3"  expression => "pass2";
+      "pass2"  expression => "pass1";
+      "pass1"  expression => "any";
+
+  methods:
+      "check"  usebundle => condition_from_command("file_copy_from_local_source_with_check_${canonified_file}_check", "${check_command} \"${source}\"", "${rc_ok}", "999"),
```
Comment (ncharles) [inline tree/30_generic_methods/file_copy_from_local_source_with_check.cf]: same here, we want the condition to be run only if the class agent is set (meaning, only if agent is running, not cf-promises, or any other ocmponent)
that would avoid triggering unwanted actions that might have side effet
(on a side note, it may be relevant to add this in the generic methods that runs actions also)

File: tests/acceptance/30_generic_methods/file_copy_from_local_source_with_check.cf [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    "destination_file_4"       string => "${tmp}/destination_test4";
+    "destination_file_canon_1" string => canonify("${destination_file_1}");
+    "destination_file_canon_2" string => canonify("${destination_file_2}");
+    "destination_file_canon_3" string => canonify("${destination_file_3}");
+    "destination_file_canon_4" string => canonify("${destination_file_4}");
+
+  files:
+    "${source_file}"
+      create => "true";
+    "${destination_file_3}"
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/file_copy_from_local_source_with_check.cf]: could you also do a test with non empty file ?

File: tree/30_generic_methods/file_copy_from_local_source_with_check.cf (hunk 2 above)
Comment (amousset) [reply tree/30_generic_methods/file_copy_from_local_source_with_check.cf]: IIRC `commands` promises are not executed by cf-promises, only `returnszero` and `execresult`

File: tree/30_generic_methods/file_copy_from_local_source_with_check.cf (hunk 1 above)
Comment (amousset) [reply tree/30_generic_methods/file_copy_from_local_source_with_check.cf]: yes!

**Comment by amousset** [comment]
Commit modified

File: tree/30_generic_methods/file_copy_from_local_source_with_check.cf (hunk 2 above)
Comment (ncharles) [reply tree/30_generic_methods/file_copy_from_local_source_with_check.cf]: you are totally right

**Comment by amousset** [comment]
I'll add an example to the doc and a contraint to allow empty code

File: tree/30_generic_methods/file_copy_from_local_source_with_check.cf [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      "class_prefix"      string => canonify(join("_", "promisers"));
+      "args"              slist => { "${source}", "${destination}", "${check_command}" };
+      "canonified_file"   string => canonify("${destination}");
+
+  defaults:
+      "rc_ok"             string => "0", if_match_regex => "";
+
+  classes:
+    agent::
+      "copy_needed" not => returnszero("diff \"${source}\" \"${destination}\"", "useshell");
```
Comment (ncharles) [inline tree/30_generic_methods/file_copy_from_local_source_with_check.cf]: could you rather use ${paths.path[diff]} , in case PATH is not set

**Comment by amousset** [comment]
Commit modified

File: tree/30_generic_methods/file_copy_from_local_source_with_check.cf [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+# * If it is one of the `rc_ok` codes, it will copy `/tmp/program.conf` into `/etc/program.conf` and return a repaired
+# * If not, it will return an error
+#
+# @parameter source      Source file (absolute path on the target node)
+# @parameter destination Destination file (absolute path on the target node)
+# @parameter check_command   Command to run, it will get the source path as argument
+# @parameter rc_ok       Return codes to be considered as valid, separated by a comma (default is 0)
+# @parameter_constraint  rc_ok "allow_empty_string" : true
+# 
+# @class_prefix file_copy_from_local_source
```
Comment (ncharles) [inline tree/30_generic_methods/file_copy_from_local_source_with_check.cf]: it should be file_copy_from_local_source_with_check

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
OK, merging this PR

**Comment by amousset** [comment]
OK, merging this PR
